File: Python/misc/bst.py
# Binary Search Tree

class Node(object):
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class BinarySearchTree(object):
    def __init__(self, root=None):
        self.root = root

    def insert(self, data):
        self.root = self._insert_util(self.root, data)

    def search(self, data):
        return self._search_util(self.root, data)

    def inorder_traversal(self, nodes):
        self._inorder_traversal_util(self.root, nodes)

    def _insert_util(self, node, data):
        if node is None:
            return Node(data)
        else:
            if data < node.data:
                node.left = self._insert_util(node.left, data)
            elif data > node.data:
                node.right = self._insert_util(node.right, data)
            return node

    def _search_util(self, node, data):
        if node is None or data is None:
            return False

        if data == node.data:
            return True
        elif data < node.data:
            return self._search_util(node.left, data)
        else:
            return self._search_util(node.right, data)

    def _inorder_traversal_util(self, node, nodes):
        if node is None:
            return
        else:
            self._inorder_traversal_util(node.left, nodes)
            nodes.append(node.data)
            self._inorder_traversal_util(node.right, nodes)


def main():
    nodes = []
    numbers = [5, 2, 1, 13, 13, 8, 4, 14, 37, 89, 89, 244, 5677]
    bst = BinarySearchTree()

    for n in numbers:
        bst.insert(n)

    bst.inorder_traversal(nodes)

    print(numbers)
    print(nodes)

    assert (bst.search(5) is True)
    assert (bst.search(2) is True)
    assert (bst.search(1) is True)
    assert (bst.search(13) is True)
    assert (bst.search(17) is False)


if __name__ == '__main__':
    main()

File: C++/misc/bst.cpp
// Binary Search Tree

#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

template <class T> struct Node {
    T data;
    Node *left;
    Node *right;
};

template <class T> class BinarySearchTree {
  public:
    BinarySearchTree(Node<T> *root) : root(root) {}
    void inorder_traversal(vector<Node<T> *> &v);
    void insert(const T &data);
    bool search(const T &data);

    friend ostream &operator<<(ostream &out, BinarySearchTree<T> &bst) {
        vector<Node<T> *> nodes;
        bst.inorder_traversal(nodes);

        for (const auto &node : nodes) {
            cout << node->data << " ";
        }

        return (out);
    }

    ~BinarySearchTree();

  private:
    Node<T> *root;
    void insert_util(Node<T> *root, const T &data);
    bool search_util(Node<T> *root, const T &data);
    void inorder_traversal_util(Node<T> *root, vector<Node<T> *> &v);
};

template <class T>
void BinarySearchTree<T>::inorder_traversal(vector<Node<T> *> &v) {
    inorder_traversal_util(root, v);
}

template <class T>
void BinarySearchTree<T>::inorder_traversal_util(Node<T> *root,
                                                 vector<Node<T> *> &v) {
    if (root) {
        inorder_traversal_util(root->left, v);
        v.push_back(root);
        inorder_traversal_util(root->right, v);
    }
}

template <class T> void BinarySearchTree<T>::insert(const T &data) {
    try {
        insert_util(root, data);
    } catch (exception &e) {
        cout << e.what() << endl;
    }
}

template <class T>
void BinarySearchTree<T>::insert_util(Node<T> *root, const T &data) {
    if (root) {
        Node<int> *node = new Node<int>;

        node->data = data;
        node->left = nullptr;
        node->right = nullptr;

        if (data < root->data) {
            if (root->left) {
                insert_util(root->left, data);
            } else {
                root->left = node;
            }
        } else if (data > root->data) {
            if (root->right) {
                insert_util(root->right, data);
            } else {
                root->right = node;
            }
        } else {
            throw runtime_error("Duplicate data not allowed in BST");
        }
    }
}

template <class T> bool BinarySearchTree<T>::search(const T &data) {
    return (search_util(root, data));
}

template <class T>
bool BinarySearchTree<T>::search_util(Node<T> *root, const T &data) {
    if (root) {
        if (data == root->data) {
            return (true);
        } else if (data < root->data) {
            return (search_util(root->left, data));
        } else {
            return (search_util(root->right, data));
        }
    } else {
        return (false);
    }
}

template <class T> BinarySearchTree<T>::~BinarySearchTree(void) {
    vector<Node<T> *> nodes;

    inorder_traversal(nodes);

    for (const auto &node : nodes) {
        delete node;
    }
}

int main(int argc, char **argv) {
    vector<Node<int> *> nodes;
    Node<int> *root = new Node<int>;
    vector<int> v = {17, 34, 8, 3567, 8, 425, 5, 3, 100, 45, 34, 23, 67};

    root->data = 10;
    root->left = nullptr;
    root->right = nullptr;

    BinarySearchTree<int> bst = BinarySearchTree<int>(root);

    for (const auto &i : v) {
        bst.insert(i);
    }

    bst.inorder_traversal(nodes);

    for (const auto &node : nodes) {
        cout << node->data << " ";
    }
    cout << endl;

    cout << bst << endl;

    assert(bst.search(425));
    assert(!bst.search(1425));

    return 0;
}
